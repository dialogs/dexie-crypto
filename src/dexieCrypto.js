// @flow strict

import Dexie, {
  type DBCore,
  type DBCoreCursor,
  type DBCoreGetManyRequest,
  type DBCoreGetRequest,
  type DBCoreMutateRequest,
  type DBCoreMutateResponse,
  type DBCoreOpenCursorRequest,
  type DBCoreQueryRequest,
  type DBCoreQueryResponse,
  type DBCoreTable,
  type DBCoreTableSchema,
  type DBCoreTransaction,
  type Middleware,
  type PromiseExtended,
} from 'dexie';
import pick from 'lodash-es/pick';
import { keepTransactionUntil } from './keepTransactionUntil';

export type CryptoCipher = {
  encrypt(data: BufferSource): Promise<ArrayBuffer>,
  decrypt(data: BufferSource): Promise<ArrayBuffer>,
};

export type DataSerializer = {
  serialize<T: { ... }>(value: T): Promise<Uint8Array>,
  deserialize<T: { ... }>(buffer: Uint8Array | ArrayBuffer): Promise<T>,
};

export function createJsonDataSerializer(): DataSerializer {
  const textEncoder = new TextEncoder();
  const textDecoder = new TextDecoder(undefined, { fatal: true });

  return {
    async serialize<T: { ... }>(value: T): Promise<Uint8Array> {
      return textEncoder.encode(JSON.stringify(value));
    },

    async deserialize<T: { ... }>(buffer): Promise<T> {
      return JSON.parse(textDecoder.decode(buffer));
    },
  };
}

export type DexieCryptoOptions = {
  /**
   * Tables for exclusion, they will be skipped by the crypto middleware.
   */
  +excludedTables?: Array<string>,
  +serializer?: DataSerializer,
};

export function createDexieCrypto(
  cipher: CryptoCipher,
  options?: DexieCryptoOptions,
): Middleware<DBCore> {
  type EncryptedValue<T: { ... }> = T & { _crypto?: ArrayBuffer, ... };

  const excludedTables: Set<string> = new Set(options?.excludedTables ?? []);
  const serializer: DataSerializer =
    options?.serializer ?? createJsonDataSerializer();

  const cachedKeyPaths: Map<string, Array<string>> = new Map();

  function getKeyPaths(schema: DBCoreTableSchema): Array<string> {
    let paths = cachedKeyPaths.get(schema.name);
    if (!paths) {
      paths = getTableKeyPaths(schema);
      cachedKeyPaths.set(schema.name, paths);
    }
    return paths;
  }

  function encrypt<T: { ... }>(
    entry: T,
    schema: DBCoreTableSchema,
  ): Promise<EncryptedValue<T>> {
    const keyPaths = getKeyPaths(schema);
    const result = pick(entry, keyPaths);

    return serializer
      .serialize(entry)
      .then((binaryData) => cipher.encrypt(binaryData))
      .then((encryptedData: ArrayBuffer) => {
        result._crypto = encryptedData;
        return result;
      });
  }

  function decrypt<T: { ... }>(encryptedValue: EncryptedValue<T>): Promise<T> {
    if (!encryptedValue._crypto) {
      return Promise.resolve(encryptedValue);
    }

    return cipher
      .decrypt(encryptedValue._crypto)
      .then((binaryData) => serializer.deserialize(binaryData));
  }

  function encryptValues<T: { ... }>(
    values: Array<T>,
    schema: DBCoreTableSchema,
  ): Promise<Array<EncryptedValue<T>>> {
    return values.length === 0
      ? Promise.resolve([])
      : Promise.all(values.map((value) => encrypt(value, schema)));
  }

  function decryptValue<T: { ... }>(
    encryptedValue: EncryptedValue<T> | null | void,
  ): Promise<T | null | void> {
    return encryptedValue
      ? decrypt(encryptedValue)
      : Promise.resolve(encryptedValue);
  }

  function decryptValues<T: { ... }>(
    encryptedValues: Array<EncryptedValue<T> | null | void>,
  ): Promise<Array<T | null | void>> {
    return encryptedValues.length === 0
      ? Promise.resolve([])
      : // $FlowFixMe
        Promise.all(encryptedValues.map((value) => decryptValue(value)));
  }

  function decryptQueryResponse(
    response: DBCoreQueryResponse,
  ): Promise<DBCoreQueryResponse> {
    return decryptValues(response.result).then((values) => ({
      ...response,
      result: values,
    }));
  }

  function createCryptoCursor(
    transaction: DBCoreTransaction,
    cursor: DBCoreCursor,
  ): DBCoreCursor {
    let decryptedValue;

    const cryptoCursor = Object.create(cursor, {
      start: {
        value: (onNext) => {
          return cursor
            .start(() => {
              keepTransactionUntil(transaction, decryptValue(cursor.value))
                .then((value) => {
                  decryptedValue = value;
                  onNext();
                  return undefined;
                })
                .catch((e) => {
                  decryptedValue = undefined;
                  cursor.fail(e);
                });
            })
            .then(() => decryptValue);
        },
      },
      value: {
        get() {
          return decryptedValue;
        },
      },
      key: {
        get() {
          // Workaround for using cursor by Dexie internal implementation.
          return cursor.key;
        },
      },
      primaryKey: {
        get() {
          // Workaround for using cursor by Dexie internal implementation.
          return cursor.primaryKey;
        },
      },
    });

    return cryptoCursor;
  }

  const middleware: Middleware<DBCore> = {
    stack: 'dbcore',
    name: 'Dexie Crypto Middleware',
    create(lowerDbCore: DBCore) {
      return {
        ...lowerDbCore,

        table(tableName) {
          const lowerTable: DBCoreTable = lowerDbCore.table(tableName);

          if (excludedTables.has(tableName)) {
            return lowerTable;
          }

          return {
            ...lowerTable,

            mutate(
              req: DBCoreMutateRequest,
            ): PromiseExtended<DBCoreMutateResponse> {
              if (req.type === 'add' || req.type === 'put') {
                return Dexie.waitFor(
                  encryptValues(req.values, lowerTable.schema),
                ).then((encryptedValues) => {
                  // $FlowFixMe
                  return lowerTable.mutate({
                    ...req,
                    values: encryptedValues,
                  });
                });
              } else {
                return lowerTable.mutate(req);
              }
            },

            get(request: DBCoreGetRequest) {
              return lowerTable
                .get(request)
                .then((encryptedValue) =>
                  Dexie.waitFor(decryptValue(encryptedValue)),
                );
            },

            getMany(
              request: DBCoreGetManyRequest,
            ): PromiseExtended<Array<mixed>> {
              return lowerTable.getMany(request).then((encryptedValues) =>
                // $FlowFixMe
                Dexie.waitFor(decryptValues(encryptedValues)),
              );
            },

            query(
              req: DBCoreQueryRequest,
            ): PromiseExtended<DBCoreQueryResponse> {
              return lowerTable.query(req).then((response) =>
                // $FlowFixMe
                Dexie.waitFor(decryptQueryResponse(response)),
              );
            },

            openCursor(
              req: DBCoreOpenCursorRequest,
            ): PromiseExtended<DBCoreCursor | null> {
              return lowerTable
                .openCursor(req)
                .then((cursor) =>
                  cursor ? createCryptoCursor(req.trans, cursor) : cursor,
                );
            },
          };
        },
      };
    },
  };

  return middleware;
}

function getTableKeyPaths(schema: DBCoreTableSchema): Array<string> {
  const resultPaths: Array<string> = [];

  [schema.primaryKey, ...schema.indexes].forEach(({ keyPath }) => {
    if (Array.isArray(keyPath)) {
      keyPath.forEach((path) => resultPaths.push(path));
    } else if (keyPath) {
      resultPaths.push(keyPath);
    }
  });

  return resultPaths;
}
