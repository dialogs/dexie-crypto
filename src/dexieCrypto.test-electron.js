// Run unit tests in Electron environment
import './dexieCrypto.test';
