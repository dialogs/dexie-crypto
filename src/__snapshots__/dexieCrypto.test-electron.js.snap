// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DexieCrypto should write encrypted data 1`] = `
Uint8Array [
  113,
  74,
  224,
  0,
  121,
  89,
  163,
  133,
  136,
  169,
  137,
  130,
  52,
  218,
  230,
  23,
  189,
  135,
  147,
  76,
  48,
  190,
  244,
  177,
  253,
  195,
  24,
  201,
  145,
  190,
  225,
  229,
  71,
  138,
  213,
  129,
  17,
  9,
  82,
  7,
  103,
  6,
  229,
  54,
  123,
  173,
  226,
  96,
  10,
  132,
  225,
  198,
  235,
]
`;
